### DISCUSSION TITLE
Failed to install realsense drivers

### DISCUSSION THREAD

**Discussion by josieammer** [prl-mushr/mushr#37] [Software]
Hi there. We are having trouble installing the realsense drivers.  

The Jetson nano was unable to boot using the mushr image. 
We followed the instructions here: https://mushr.io/hardware/nano_install/
The build fails to install the realsense drivers. 

mushr_install_hw_drivers.sh fails at the following line:
sudo add-apt-repository "deb http://realsense-hw-public ...

starting back at Intel's instructions: https://github.com/IntelRealSense/librealsense/blob/development/doc/distribution_linux.md
It suggests a different server. 
sudo add-apt-repository "deb https://librealsense.intel.com/Debian/apt-repo $(lsb_release -cs) main" -u

But that command fails with the same error.  

E: Failed to fetch http://realsense-hw-public.s3.amazonaws.com/Debian/apt-repo/dists/bionic/InRelease  403  Forbidden [IP: 52.218.21.217 80]
E: The repository 'http://realsense-hw-public.s3.amazonaws.com/Debian/apt-repo bionic InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.

Seems similar to this issues here:
https://github.com/IntelRealSense/librealsense/issues/1670
That thread seems to suggest I build from scratch.  Before I do that, I was wondering if you had a different workaround for this error.

**Comment by josieammer** [comment]
Update: 
We successfully followed the instructions here to rebuild librealsense from scratch: 
https://github.com/IntelRealSense/librealsense/blob/development/doc/installation.md

At step 4, the right script to use for the Jetson Nano is:
./scripts/patch-realsense-ubuntu-L4T.sh

We are able to open realsense-viewer and connect to the camera.

**Reply by palanc** [reply]
Hi @josieammer - Thanks for letting us know that the scripts are no longer working! I'm glad that you were able to build it from scratch. In case that ends up not working for you, I believe that a combination of updating the URL repository to https://librealsense.intel.com... (as you mentioned in your original message) and adding the receive key from this site to be trusted should work. In other words:

`sudo add-apt-repository "deb https://librealsense.intel.com/Debian/apt-repo $(lsb_release -cs) main" -u`
`sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys F42ED6FBAB17C654`

We have updated our scripts to reflect this change, and are currently testing them to make sure that they work. I will reply again here once it has been merged into our main branch.

**Reply by palanc** [reply]
This fix has been merged into our main branch. Thanks again for pointing out the issue!

**Reply by Lilly-yang** [reply]
Hi @planc509,

Could you please tell me how do you update the Realsense SDK? I run the latest [mushr_install.sh](https://github.com/prl-mushr/mushr/blob/master/mushr_utils/install_scripts/mushr_install_hw_drivers.sh) file on Jetson Nano (installed MuSHR image) but the Realsense SDK still with version v2.24.0, and the suggested firmware Signed_Image_UVC_5_11_6_250 is no longer support D345i camera (Minimum Version for D435i is 5.12.07.100). But when I run it on my own laptop, which without Realsense SDK, it installed the latest version of Realsense SDK: v2.47.0.

I have this question because it's kind of complicated to uninstall and reinstall Realsens SDK manually on Jetson Nano. I have met a bunch of errors. And also I have already upgraded the cameras' firmware to the latest version on my laptop too. And somehow I can't downgrade it. I met some USB input/output problems when I use v2.24 SDK on Jetson Nano. I can't downgrade it on my laptop with the latest version(v2.47) of Realsense SDK neither (it doesn't match the old firmware). And as I know we can only downgrade the camera 20 times. after that, we can only upgrade it.

With the wrong version of realsense-ros, Realsense SDK, and camera firmware, I always receive a 'there is no data' error when n I launch the camera. I would appreciate you can give me any suggestions to help me run the camera properly with all data(topics) that can be reached.

**Reply by palanc** [reply]
Hi @Lilly-yang - I'm not able to completely replicate the state of your Nano, so we might have to try a few things. Here's what I'd suggest first:

1. **Remove the defunct apt repo:**
    Open the file **/etc/apt/sources.list** with sudo privileges
    Remove the line that says `deb http://realsense-hw-public.s3.amazonaws.com/Debian/apt-repo bionic main` 
2. **Get latest version of librealsense:**
    `sudo apt-key adv --keyserver keys.gnupg.net --recv-key F6E65AC044F831AC80A06380C8B3A55A6F3EFCDE || sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-key F6E65AC044F831AC80A06380C8B3A55A6F3EFCDE`
    `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys F42ED6FBAB17C654`
    `sudo add-apt-repository "deb https://librealsense.intel.com/Debian/apt-repo $(lsb_release -cs) main" -u`
    `sudo apt update`
    `sudo apt-get install apt-utils`
    `sudo apt-get install librealsense2-utils librealsense2-dev -y`
3. **Verify install of updated librealsense:**
    `realsense-viewer`
    At the top of the window that opens, verify that it reads _Intel RealSense Viewer **v2.41.0**_
4. **Rebuild realsense2_camera: (here I'm assuming your workspace is named catkin_ws)**
    `cd ~/catkin_ws`
    `rm build/mushr/mushr_hardware/realsense/realsense2_camera devel/include/realsense2_camera` 
    `rm devel/lib/librealsense2_camera.so devel/share/realsense2_camera`
    `catkin_make`
5. **Test realsense2_camera:**
    `roslaunch realsense2_camera rs_camera.launch`
    Then use either `rostopic` or `rviz` to verify that the camera is publishing data

Try that out and let us know how it goes.

**Reply by Lilly-yang** [reply]
Hi @planc509,

Thank you for your detailed reply. I found the reason why I cannot upgrade the Realsense SDK. The Realsense SDK in the Mushr image was built from the source. So it cannot be deleted by the commands you gave above. The way to delete it is with `sudo make uninstall` invoked from the Librealsense build directory (`/home/librealsense/build`). I found the solution [here](https://github.com/IntelRealSense/librealsense/issues/5773#issuecomment-581477785).

After successfully deleting the Realsense SDK that was built from the source, I can install, upgrade, and uninstall the Realsense SDK according to the method you provided.

**Reply by Lilly-yang** [reply]
And no problem to launch two cameras (D453i and T265).
![viewer_2_camera](https://user-images.githubusercontent.com/39558023/123825744-753b1b00-d8ff-11eb-8086-f5bb73795036.png)

More information:
```
RealSense ROS v2.2.21
LibRealSense v2.41.0
Intel RealSense D435I - FW version: 05.12.10.00
Intel RealSense T265 - FW version: 0.2.0.951
```

**Reply by palanc** [reply]
I see, sorry I had assumed you installed it using the install script. Glad you got it figured out!
